pkg/core/region.go

```diff
@@ -1826,37 +1827,86 @@ func (r *RegionsInfo) ScanRegions(startKey, endKey []byte, limit int) []*RegionI
 // BatchScanRegions scans regions in given key pairs, returns at most `limit` regions.
 // limit <= 0 means no limit.
 // The given key pairs should be non-overlapping.
-func (r *RegionsInfo) BatchScanRegions(keyRanges *KeyRanges, limit int) []*RegionInfo {
-	r.t.RLock()
-	defer r.t.RUnlock()
-
+func (r *RegionsInfo) BatchScanRegions(keyRanges *KeyRanges, opts ...BatchScanRegionsOptionFunc) ([]*RegionInfo, error) {
+	keyRanges.Merge()
 	krs := keyRanges.Ranges()
 	res := make([]*RegionInfo, 0, len(krs))
-	var lastRegion *RegionInfo
+
+	scanOptions := &batchScanRegionsOptions{}
+	for _, opt := range opts {
+		opt(scanOptions)
+	}
+
+	r.t.RLock()
+	defer r.t.RUnlock()
 	for _, keyRange := range krs {
-		if limit > 0 && len(res) >= limit {
-			return res
+		if scanOptions.limit > 0 && len(res) >= scanOptions.limit {
+			res = res[:scanOptions.limit]
+			return res, nil
 		}
-		if lastRegion != nil {
-			if lastRegion.Contains(keyRange.EndKey) {
-				continue
-			} else if lastRegion.Contains(keyRange.StartKey) {
-				keyRange.StartKey = lastRegion.GetEndKey()
+
+		regions, err := scanRegion(r.tree, keyRange, scanOptions.limit)
+		if err != nil {
+			log.Warn("scan regions failed", zap.Bool("outputMustContainAllKeyRange",
+				scanOptions.outputMustContainAllKeyRange), zap.Error(err))
+			if scanOptions.outputMustContainAllKeyRange {
+				return nil, err
 			}
 		}
-		r.tree.scanRange(keyRange.StartKey, func(region *RegionInfo) bool {
-			if len(keyRange.EndKey) > 0 && bytes.Compare(region.GetStartKey(), keyRange.EndKey) >= 0 {
-				return false
-			}
-			if limit > 0 && len(res) >= limit {
-				return false
-			}
-			lastRegion = region
-			res = append(res, region)
-			return true
-		})
+		if len(res) > 0 && len(regions) > 0 && res[len(res)-1].meta.Id == regions[0].meta.Id {
+			// skip the region that has been scanned
+			regions = regions[1:]
+		}
+		res = append(res, regions...)
 	}
-	return res
+	return res, nil
+}
+
+func scanRegion(regionTree *regionTree, keyRange *KeyRange, limit int) ([]*RegionInfo, error) {
+	var (
+		res        []*RegionInfo
+		lastRegion = &RegionInfo{
+			meta: &metapb.Region{EndKey: keyRange.StartKey},
+		}
+		exceedLimit = func() bool { return limit > 0 && len(res) >= limit }
+		err         error
+	)
+	regionTree.scanRange(keyRange.StartKey, func(region *RegionInfo) bool {
+		if len(keyRange.EndKey) > 0 && len(region.GetStartKey()) > 0 &&
+			bytes.Compare(region.GetStartKey(), keyRange.EndKey) >= 0 {
+			return false
+		}
+		if exceedLimit() {
+			return false
+		}
+		if len(lastRegion.GetEndKey()) > 0 && len(region.GetStartKey()) > 0 &&
+			bytes.Compare(region.GetStartKey(), lastRegion.GetEndKey()) > 0 {
+			err = multierr.Append(err, errs.ErrRegionNotAdjacent.FastGen(
```

Comment:
Can this function accept parameter `outputMustContainAllKeyRange` and fast fail and return here?

Reply:
I will try to add it~

Reply:
I added the fast fail mechanism. PTAL again  @you06 cc @lance6716 
https://github.com/tikv/pd/pull/8375/commits/53f1e62aa3d55cedced9df810932d2485ef347f1